internal/babysitter/babysitter.go

```diff
@@ -254,50 +264,88 @@ func (b *Babysitter) ActivateComponent(ctx context.Context, req *protos.Activate
+				continue
+			}
+			break
+		}
+		return &protos.ActivateComponentReply{}, nil
+	}
 	b.mu.Lock()
 	defer b.mu.Unlock()
 	if b.watchingRoutingInfo == nil {
 		b.watchingRoutingInfo = map[string]struct{}{}
 	}
 	if _, ok := b.watchingRoutingInfo[req.Component]; !ok {
 		b.watchingRoutingInfo[req.Component] = struct{}{}
-		go b.watchRoutingInfo(req.Component)
+		if req.Routed {
+			go b.watchRouted(b.ctx, req.Component)
+		} else {
+			go b.watchNonRouted(b.ctx, req.Component, targetReplicaSet)
+		}
 	}
 	return &protos.ActivateComponentReply{}, nil
 }
 
-func (b *Babysitter) watchRoutingInfo(component string) {
+// watchRouted watches and updates the routing information for a routed
+// component.
+func (b *Babysitter) watchRouted(ctx context.Context, component string) {
 	version := ""
-	for r := retry.Begin(); r.Continue(b.ctx); {
-		routing, newVersion, err := b.getRoutingInfo(component, version)
+	for r := retry.Begin(); r.Continue(ctx); {
+		reply, err := b.manager.GetRoutingInfo(ctx, &nanny.GetRoutingRequest{
+			Component: component,
+			Version:   version,
+			Config:    b.cfg,
+		})
 		if err != nil {
-			b.logger.Error("cannot get routing info; will retry", "err", err, "component", component)
+			b.logger.Error("cannot get routing info for a routed component; will retry", "err", err, "component", component)
 			continue
 		}
-		if err := b.envelope.UpdateRoutingInfo(routing); err != nil {
-			b.logger.Error("cannot update routing info; will retry", "err", err, "component", component)
+		if err := b.envelope.UpdateRoutingInfo(reply.Routing); err != nil {
+			b.logger.Error("cannot update routing info for a routed component; will retry", "err", err, "component", component)
 			continue
 		}
-		version = newVersion
-		if routing.Local {
-			// If the routing is local, it will never change. There is no need
-			// to watch.
-			return
-		}
+		version = reply.Version
 		r.Reset()
 	}
 }
 
-func (b *Babysitter) getRoutingInfo(component string, version string) (*protos.RoutingInfo, string, error) {
-	reply, err := b.manager.GetRoutingInfo(b.ctx, &nanny.GetRoutingRequest{
-		Component: component,
-		Version:   version,
-		Config:    b.cfg,
-	})
-	if err != nil {
-		return nil, "", err
+// watchRouted watches and updates the routing information for an component
+// that isn't routed.
+func (b *Babysitter) watchNonRouted(ctx context.Context, component, replicaSet string) {
+outer:
+	for r := retry.Begin(); r.Continue(ctx); {
+		watcher, err := b.newReplicaWatcher(ctx, replicaSet)
+		if err != nil {
+			b.logger.Error("cannot create replica watcher for an unrouted component; will retry", "err", err, "component", component)
+			continue
+		}
+		for {
+			replicas, err := watcher.GetReplicas(ctx)
+			if err != nil {
+				b.logger.Error("cannot get replicas for an unrouted component; will retry", "err", err, "component", component)
+				continue outer
+			}
+			r.Reset()
+			if err := b.envelope.UpdateRoutingInfo(&protos.RoutingInfo{
```

Comment:
One difference between `watchNonRouted` and `watchRouted` is that `watchRouted` retries failed `UpdateRoutingInfo` calls with the same routing info, whereas `watchNonRouted` will retry with a newer routing info. Looking at the weavelet `UpdateRoutingInfo` code, it looks all errors are not retryable, so maybe retrying with a newer routing info makes more sense. Feel free to do nothing for now and we can fix later when we think harder about failures and retries.